### PR TITLE
Add wildcard tags json parser support

### LINKED ISSUE #7531: Json Parser wildcard tagkey support
## Feature Request

Opening a feature request kicks off a discussion on json parser to support wildcard tag key. Discussion initially happens on [here](https://influxcommunity.slack.com/archives/CH99HUH8V/p1589474801422000).

### Proposal:
Allow Parser to handle TagKey creation automatically. 

### Current behavior:
The input json parser doesnt support output from [json serializers](https://github.com/influxdata/telegraf/tree/master/plugins/serializers/json) properly.

Telegraf Setting
```
[[inputs.kafka_consumer]]
   brokers = ["xxx:9094"]
   topics = ["dev-test"]
   consumer_group = "host"
   insecure_skip_verify = true
   data_format = "json"
   json_name_key = "name"
   json_time_key = "timestamp"
   json_time_format = "unix"
   json_strict = true
   json_string_fields = []
   tag_keys = ["tags"]
   offset = "oldest"
```

Input Json:
```
{
    "fields": {
        "value": 2
    },
    "name": "metrics_name.test",
    "tags": {
        "InstanceId": "i-123",
        "app": "test",
        "availability-zone": "ap-southeast-1z",
        "cluster": "dev",
        "env": "development",
        "host": "ip-10.compute.internal",
        "metric_type": "counter"
    },
    "timestamp": 1589376103
}

```
The metrics output tend to skip all the tag unless it is specified in `tag_keys = ["tags_InstanceId", "tags_app", , "tags_availability-zone", "tags_cluster", "tags_env", "tags_host", "tags_metric_type"]`

This tend to be inconvenience as configuration need to be updated when new tag are insert. Since the output from `serializers` formatted this way, parser should made compatible to each other. 

### Desired behavior:

A new config is probably suggested or `*` should be allow in `tag_keys`.

```
[[inputs.kafka_consumer]]
   ...
   data_format = "json"
   json_name_key = "name"
   json_time_key = "timestamp"
   json_time_format = "unix"
   json_strict = true
   json_string_fields = []
   tag_keys = ["tags_*"]
```

### Use case: 

This tend to be inconvenience as configuration need to be updated when new tag are insert. Since the output from `serializers` formatted this way, parser should made compatible to each other.

### PR DESCRIPTION
**Pull request by helenosheaa**
Closes #7531

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,4 +4,4 @@
 /telegraf.exe
 /telegraf.gz
 /vendor
-.DS_Store
+.DS_Store
```
Comment (srebhan) [inline .gitignore:7]: This is unrelated, please revert. We really want to have a newline there. :-)

File: plugins/parsers/json/README.md [hunk 2]
```diff
@@ -30,10 +30,12 @@ ignored unless specified in the `tag_key` or `json_string_fields` options.
   json_query = ""
 
   ## Tag keys is an array of keys that should be added as tags.  Matching keys
-  ## are no longer saved as fields.
+  ## are no longer saved as fields. Now supports wildcard glob matching.
```
Comment (srebhan) [inline plugins/parsers/json/README.md:33]: Maybe leave out the "Now" as this might be outdated in a few month.

File: plugins/parsers/json/parser.go [hunk 3]
```diff
@@ -49,13 +49,20 @@ type Parser struct {
 
 func New(config *Config) (*Parser, error) {
 	stringFilter, err := filter.Compile(config.StringFields)
+
```
Comment (srebhan) [inline plugins/parsers/json/parser.go:52]: Skip newline

File: plugins/parsers/json/parser.go [hunk 4]
```diff
@@ -49,13 +49,20 @@ type Parser struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	tagKeyFilter, err := filter.Compile(config.TagKeys)
+
```
Comment (srebhan) [inline plugins/parsers/json/parser.go:58]: Same here, skip newline.

File: plugins/parsers/json/parser.go [hunk 5]
```diff
@@ -150,27 +157,32 @@ func (p *Parser) parseObject(data map[string]interface{}, timestamp time.Time) (
...
-			tags[name] = strconv.FormatFloat(value, 'f', -1, 64)
+			tags[name] = strconv.FormatFloat(t, 'f', -1, 64)
 			delete(fields, name)
 		default:
 			log.Printf("E! [parsers.json] Unrecognized type %T", value)
 		}
 	}
 
-	//remove any additional string/bool values from fields
+	// remove any additional string/bool values from fields
```
Comment (srebhan) [inline plugins/parsers/json/parser.go:185]: Remove whitespace to keep format similar to all other comments...

File: plugins/parsers/json/parser_test.go [hunk 6]
```diff
@@ -27,7 +27,11 @@ const validJSONTags = `
         "c": 6
     },
     "mytag": "foobar",
-    "othertag": "baz"
+	"othertag": "baz",
+	"tags_object": {
+		"mytag": "foobar",
+		"othertag": "baz"
+	}
```
Comment (srebhan) [inline plugins/parsers/json/parser_test.go:34]: There are some formatting issues. Please run `make fmt`!

File: plugins/parsers/json/parser_test.go [hunk 7]
```diff
@@ -948,3 +970,304 @@ func TestParse(t *testing.T) {
...
+	require.Len(t, metrics, 1)
+	require.Equal(t, "json_test", metrics[0].Name())
+	require.Equal(t, map[string]interface{}{
+		"a":   float64(5),
+		"b_c": float64(6),
+	}, metrics[0].Fields())
+	require.Equal(t, map[string]string{
+		"tags_object_mytag":    "foobar",
+		"tags_object_othertag": "baz",
+	}, metrics[0].Tags())
```
Comment (srebhan) [inline plugins/parsers/json/parser_test.go:992]: Please change this to a config-expected table approach. That is, define a new struct e.g.
```
var tests = []struct {
		config       []string
		expected  telegraf.Metric
	}{
		[]string{"tags_object_*"},
		testutil.MustMetric(
			"json_test",
			map[string]string{
				"tags_object_mytag":    "foobar",
				"tags_object_othertag": "baz",
			},
			map[string]interface{}{
￼				"a":   float64(5),
￼		￼		"b_c": float64(6),
￼			},
			time.Unix(0, 0),
		)
	},
	{
	...
	},
```
and then iterate over this table running your tests. You can even compare the resulting metrics using`testutil.RequireMetricEqual()`.

**Review by srebhan — CHANGES_REQUESTED** [review]
Minor (whitespace) comments in the code. However, the test can be simplified using a table approach defining the config and the expected outcome and then do the logic only once in a for loop over that table.

File: plugins/parsers/json/parser_test.go (hunk 6 above)
Comment (ssoroka) [reply plugins/parsers/json/parser_test.go:34]: `make fmt` won't catch this as it's in a string. Please align manually. :D

File: plugins/parsers/json/parser.go (hunk 5 above)
Comment (ssoroka) [reply plugins/parsers/json/parser.go:185]: I'd prefer a single space after each `//`, for what it's worth.

Comment (srebhan) [reply plugins/parsers/json/parser.go:183]: Well then we should modify all comments here... ;-)

File: plugins/parsers/json/parser_test.go (hunk 6 above)
Comment (srebhan) [reply plugins/parsers/json/parser_test.go]: Didn't know...

**Comment by srebhan** [comment]
@helenosheaa the fixes look good, can you please also address the comment in the test-file!?

File: .gitignore (hunk 1 above)
Comment (helenosheaa) [reply .gitignore]: Reverted my bad!

File: plugins/parsers/json/README.md (hunk 2 above)
Comment (helenosheaa) [reply plugins/parsers/json/README.md]: Done

File: plugins/parsers/json/parser.go (hunk 3 above)
Comment (helenosheaa) [reply plugins/parsers/json/parser.go]: Done

File: plugins/parsers/json/parser.go (hunk 4 above)
Comment (helenosheaa) [reply plugins/parsers/json/parser.go]: Done

File: plugins/parsers/json/parser.go (hunk 5 above)
Comment (helenosheaa) [reply plugins/parsers/json/parser.go:183]: I've modified all comments to have the space for consistency

Comment (helenosheaa) [reply plugins/parsers/json/parser.go:183]: I've added spaces for all the comments to keep it consistent

File: plugins/parsers/json/parser_test.go (hunk 6 above)
Comment (helenosheaa) [reply plugins/parsers/json/parser_test.go]: Done

File: plugins/parsers/json/parser_test.go (hunk 7 above)
Comment (helenosheaa) [reply plugins/parsers/json/parser_test.go]: I've refactored to using the config-expected table approach :)

**Review by helenosheaa — COMMENTED** [review]
@srebhan yes that's the test file changed now as well, thanks for the review :)

**Review by ssoroka — APPROVED** [review]

**Review by srebhan — COMMENTED** [review]
Looks good. Thanks for the changes!
